seaborn/_core/data.py

```diff
@@ -312,8 +312,18 @@ def convert_dataframe_to_pandas(data: object) -> pd.DataFrame:
         # in Plot.add(). But noting here in case this seems to be a bottleneck.
         return pd.api.interchange.from_dataframe(data)
     except Exception as err:
-        msg = (
-            "Encountered an exception when converting data source "
-            "to a pandas DataFrame. See traceback above for details."
-        )
-        raise RuntimeError(msg) from err
+        if hasattr(data, 'to_pandas'):
+            try:
+                return data.to_pandas()  # type: ignore[attr-defined]
+            except Exception:
+                msg = (
+                    "Encountered an exception when converting data source "
+                    "to a pandas DataFrame. See traceback above for details."
+                )
+                raise RuntimeError(msg) from err
```

Comment:
Not wild about the copy-paste here, I guess there's no super clean way to write this sort of nested exception handling, but can we please at least just define `msg` once?

(I guess technically you don't even need to raise from the inner `except` block, because the outer raise statement doesn't need to be guarded by an `else`. You could either return from `to_pandas` or just `pass` on any exception that comes out of it and end up back at the original `raise` statement. Not 100% sure that is cleaner / more explicit though).